datastore/additional_data/sources/nspl.py

```diff
@@ -0,0 +1,166 @@
+                            if record[field] == 99.999999:
+                                record[field] = None
+                    if record["lat"] and record["long"]:
+                        record["location"] = {
+                            "lat": record["lat"],
+                            "lon": record["long"],
+                        }
+
+                    # integer fields
+                    for field in [
+                        "oseast1m",
+                        "osnrth1m",
+                        "usertype",
+                        "osgrdind",
+                        "imd",
+                    ]:
+                        if record[field]:
+                            record[field] = int(record[field])
+
+                    # add postcode hash
+                    record["hash"] = hashlib.md5(
+                        record["pcds"].lower().replace(" ", "").encode()
+                    ).hexdigest()
+
+                    postcodes.append(record)
+                    postcode_count += 1
+                print("[postcodes] Processed %s postcodes" % postcode_count)
+
+        return postcodes
+
+    def save_nspl_data(self, data):
+        for record in data:
+            nspl_postcode = (
+                record["pcd"]
+                if record["pcd"]
+                else record["pcd2"]
+                if record["pcd2"]
+                else record["pcds"]
+            )
+            postcode = "".join(nspl_postcode.split()).upper()
+            NSPL.objects.create(postcode=postcode, data=record).save()
+
+    def import_nspl(self, url=NSPL_URL):
+        """
+        Example of a data db entry:
+        {
+            'ccg': 'S03000012', 'ced': None, 'cty': None, 'eer': 'S15000001', 'imd': 6808, 'lat': 57.101474,
+            'pcd': 'AB1 0AA', 'pct': 'S03000012', 'pfa': 'S23000009', 'rgn': None, 'stp': None, 'ctry': 'S92000003',
+            'hash': '3b7db6a231babf449f1c6d7a61ce317f', 'laua': 'S12000033', 'lep1': None, 'lep2': None,
+            'long': -2.242851, 'nuts': 'S31000935', 'oa11': 'S00090303', 'park': None, 'pcd2': 'AB1  0AA',
+            'pcds': 'AB1 0AA', 'pcon': 'S14000002', 'ttwa': 'S22000047', 'ward': 'S13002843', 'wz11': 'S34002990',
+            'bua11': None, 'nhser': None, 'oac11': '1C3', 'calncv': None, 'dointr': '1980-01-01 00:00:00',
+            'doterm': '1996-06-01 00:00:00', 'hlthau': 'S08000020', 'lsoa11': 'S01006514', 'msoa11': 'S02001237',
+            'teclec': 'S09000001', 'buasd11': None, 'ru11ind': '3', 'location': {'lat': 57.101474, 'lon': -2.242851},
+            'oseast1m': 385386, 'osgrdind': 1, 'osnrth1m': 801193, 'usertype': 0
+        }
+        Field names information can be found in NSPL User Guide at https://geoportal.statistics.gov.uk/
+        """
+        zip_file = self.get_zipfile(url)
+        data = self.get_location_data(zip_file)
+        data_json = json.loads(json.dumps(data))
+
+        if NSPL.objects.exists():
+            NSPL.objects.all().delete()
```

Comment:
No need to check if exists first.
We ought to make the delete behaviour consistent with the load_org_data which has a separate command delete_org_data. I think there may be a use case where you want to delete all the data but not start adding it again... not sure.

Reply:
I think that `.delete()` will fail if the object doesn't exist.